### PR TITLE
[LEGACY] Hetero 2.0: Implement `hetero::CompiledModel`

### PR DESCRIPTION
**Pull request by vurusovs**
### Details:
 - `hetero::CompiledModel` part of https://github.com/openvinotoolkit/openvino/pull/17814 and https://github.com/openvinotoolkit/openvino/pull/17605/

### Tickets:
 - CVS-109770

### CODE REVIEW

File: src/inference/dev_api/openvino/runtime/icompiled_model.hpp [hunk 1]
```diff
@@ -151,11 +149,13 @@ class OPENVINO_RUNTIME_API ICompiledModel : public std::enable_shared_from_this<
     friend ov::IExecutableNetworkWrapper;
     friend InferenceEngine::ICompiledModelWrapper;
 
+protected:
+    std::vector<ov::Output<const ov::Node>> m_inputs;
+    std::vector<ov::Output<const ov::Node>> m_outputs;
```
Comment (ilyachur) [inline src/inference/dev_api/openvino/runtime/icompiled_model.hpp:154]: Why do you need this change?
get_inputs() and get_outputs() methods are virtual, you can override them

File: src/plugins/hetero/infer_request.cpp [hunk 2]
```diff
@@ -13,7 +13,7 @@
 #include <map>
 #include <string>
 
-#include "itt.hpp"
+#include "src/itt.hpp"
```
Comment (ilyachur) [inline src/plugins/hetero/infer_request.cpp:16]:
```suggestion
#include "itt.hpp"
```

File: src/plugins/hetero/src/compiled_model.cpp [hunk 3]
```diff
@@ -0,0 +1,722 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "itt.hpp"
```
Comment (ilyachur) [inline src/plugins/hetero/src/compiled_model.cpp:5]: Please fix code style

File: src/plugins/hetero/src/compiled_model.cpp [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+#include "xml_parse_utils.h"
+
+#include "ie_ngraph_utils.hpp"
+#include "ie_plugin_config.hpp"
+#include "ie_algorithm.hpp"
+
+#include "converter_utils.hpp"
+
+template <typename T>
+using NodeMap = std::unordered_map<ngraph::Node*, T>;
```
Comment (ilyachur) [inline src/plugins/hetero/src/compiled_model.cpp:34]:
```suggestion
using NodeMap = std::unordered_map<ov::Node*, T>;
```

File: src/plugins/hetero/src/compiled_model.cpp [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+        std::unordered_set<std::string> devices;
+        NodeMap<std::string> affinities;
+        // Check that all nodes has user or plugin defined affinities
+        for (auto&& node : orderedOps) {
+            auto itAffinity = queryNetworkResult.find(node->get_friendly_name());
+            if (itAffinity != queryNetworkResult.end()) {
+                affinities[node.get()] = itAffinity->second;
+                devices.emplace(itAffinity->second);
+            } else if (allEmpty) {
+                IE_THROW() << "Hetero device used default fallback policy, but some layers eg: \n(Name:"
```
Comment (ilyachur) [inline src/plugins/hetero/src/compiled_model.cpp:88]: Please throw only new exceptions
```suggestion
                OPENVINO_THROW() 
```

File: src/plugins/hetero/src/compiled_model.cpp [hunk 6]
```diff
@@ -0,0 +1,722 @@
...
+        OPENVINO_ASSERT(subgraphTopoSortsStep < subgraphs.size());
+        ++subgraphTopoSortsStep;
+        std::vector<Subgraph> newOrderedSubgraphs;
+        auto IsOrderedSubGraph = [&](const Subgraph& subgraph) {
+            auto& parameters = subgraph._parameters;
+            return std::all_of(parameters.begin(),
+                               parameters.end(),
+                               [&](const ngraph::ParameterVector::value_type& parameter) {
+                                   return InferenceEngine::details::contains(graphInputNodes, parameter.get()) ||
+                                          InferenceEngine::details::contains(prevResults, subgraphParameterToPrevResult[parameter.get()]);
```
Comment (ilyachur) [inline src/plugins/hetero/src/compiled_model.cpp:319]: Please do not use legacy namespaces

File: src/inference/dev_api/openvino/runtime/icompiled_model.hpp (hunk 1 above)
Comment (vurusovs) [reply src/inference/dev_api/openvino/runtime/icompiled_model.hpp]: Not used now

File: src/plugins/hetero/infer_request.cpp (hunk 2 above)
Comment (vurusovs) [reply src/plugins/hetero/infer_request.cpp:16]: Fixed

File: src/plugins/hetero/src/compiled_model.cpp (hunk 3 above)
Comment (vurusovs) [reply src/plugins/hetero/src/compiled_model.cpp]: Fixed

File: src/plugins/hetero/src/compiled_model.cpp (hunk 4 above)
Comment (vurusovs) [reply src/plugins/hetero/src/compiled_model.cpp:29]: Replaced

File: src/plugins/hetero/src/compiled_model.cpp (hunk 5 above)
Comment (vurusovs) [reply src/plugins/hetero/src/compiled_model.cpp:82]: Done

File: src/plugins/hetero/src/compiled_model.cpp (hunk 6 above)
Comment (vurusovs) [reply src/plugins/hetero/src/compiled_model.cpp]: Will be fixed in https://github.com/openvinotoolkit/openvino/pull/18429

**Comment by vurusovs** [comment]
Replaced with https://github.com/openvinotoolkit/openvino/pull/18392

**Pull request closed without merging**